### PR TITLE
Make generate-js.js ts clean

### PR DESCRIPTION
**Pull request by markw65**
While working on #429 I found that having @ts-check enabled on the new files was very helpful, so I decided to see how hard it would be to turn on for generate-js.js. This is the result.

I found several typing errors (either errors in jsdoc comments, or in peg.d.ts) - but these were just incorrect declarations - the code itself was fine (also, most of them were introduced by me, when I added source map support to generate-js.js).

I also found one possible bug (again, it was me that did it): A SourceNode was constructed with a child array consisting of strings and an array of SourceNodes (ie `Array<string|Array<SourceNode>>`). The type info says it should be a `Array<string|SourceNode>`. Apparently the code did work though, so who knows.

In any case, if this seems worthwhile, I'll be happy to put up similar pull requests for other files as I get time.

### CODE REVIEW

File: lib/compiler/passes/generate-js.js [hunk 1]
```diff
@@ -204,19 +269,24 @@ function generateJS(ast, options) {
...
           (c, i) => "  var " + l(i) + " = " + buildLiteral(c) + ";"
-        ).concat("", ast.classes.map(
+        ).concat("", (ast.classes || []).map(
           (c, i) => "  var " + r(i) + " = " + buildRegexp(c) + ";"
-        )).concat("", ast.expectations.map(
+        )).concat("", (ast.expectations || []).map(
           (c, i) => "  var " + e(i) + " = " + buildExpectation(c) + ";"
         )).concat("").join("\n"),
-        ast.functions.map(buildFunc),
+        ...(ast.functions || []).map(buildFunc),
```
Comment (markw65) [inline lib/compiler/passes/generate-js.js:279]: This was the bug, where SourceNode's type info says the array can contain SourceNodes and strings, but not arrays of SourceNodes.

File: lib/peg.d.ts [hunk 2]
```diff
@@ -51,6 +51,23 @@ declare namespace ast {
...
+    ignoreCase: boolean;
+  }
+
+  type Expectation =
+    | { type: "any" }
+    | { type: "literal"; value: string; ignoreCase: boolean }
+    | { type: "rule"; value: string }
+    | CharacterRange & { type: "class" }
+    ;
+
```
Comment (markw65) [inline lib/peg.d.ts:70]: @hildjj 

I don't like this change (too much confusion, and overlap between `ast.Expectation`, and `parser.Expectation`) but I think its a correct description of what's going on.

Previously, `Grammar.Expectations` was typed as `parser.Expectation[]`, but that's not right, because `parser.ClassExpectation`, has a `value` field, rather than a `parts` field.

Also, `parser.Expectation` supports types `any`, `class`, `end`, `literal`, and `other`, while `Grammar.expectations[]` supports types `any`, `class`, `literal` and `rule` - with assertions for any other types that appear.

Similarly `Grammar.classes[]` has `value`, rather than `parts`.

This seems pointlessly confusing. Essentially generate-bytecode.js takes an object with a `parts` array, and converts it to an object with a `value` array ([here](https://github.com/peggyjs/peggy/blob/main/lib/compiler/passes/generate-bytecode.js#L287-L291) and [here](https://github.com/peggyjs/peggy/blob/main/lib/compiler/passes/generate-bytecode.js#L1016-L1020). And then generate-js.js [converts it back again](https://github.com/peggyjs/peggy/blob/main/lib/compiler/passes/generate-js.js#L1009).

It would be nice to fix it so that `Grammar.classes` and `Grammar.expectations` both use `parts` rather than `values` (and maybe rename `rule` as `other`) - but presumably that could break plugins?

If we're stuck with this, I think I should rename these to something like `GrammarCharacterClass` and `GrammarExpectation` so its clear they're just for this one use...

**Comment by hildjj** [comment]
I think this is the next one I want to land.  I'll look at it later today, but could you do a quick rebase please?

**Comment by markw65** [comment]
Rebased

File: lib/peg.d.ts (hunk 2 above)
Comment (hildjj) [reply lib/peg.d.ts:70]: Let's go ahead and at least do the rename.  It will be easier to see the consolidation change after.

File: lib/compiler/passes/generate-js.js [hunk 3]
```diff
@@ -113,13 +161,37 @@ function generateJS(ast, options) {
...
+     * @returns {SourceArray}
+     */
+    /**
+     * @param {SourceArray | string | SourceNode} code
+     * @returns {SourceArray | string | SourceNode}
+     */
     function helper(code) {
       if (Array.isArray(code)) {
-        return code.map(helper);
+        return code.map(s => helper(s));
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:190]: I was having trouble typing helper correctly, and at one point I did this so I could explicitly insert a cast. But in the end I got the types right, and didn't need it. I think I can revert this.

File: lib/compiler/passes/generate-js.js [hunk 4]
```diff
@@ -113,13 +161,37 @@ function generateJS(ast, options) {
...
       }
       if (code instanceof SourceNode) {
         inSourceNode++;
-        code.children = helper(code.children);
+        code.children = (helper(code.children));
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:194]: Again, at one point I had to insert a cast: `/** @type {SourceNode[]} */(helper(code.children))`. But then I got the types right, but failed to clean up. Will fix.

File: lib/compiler/passes/generate-js.js [hunk 5]
```diff
@@ -80,9 +85,52 @@ function wrapInSourceNode(prefix, chunk, location, suffix, name) {
...
+/**
+ * Generates parser JavaScript code.
+ *
+ * @param {PEG.ast.Grammar} ast
+ * @param {Options} options
+ */
 function generateJS(ast, options) {
-  // These only indent non-empty lines to avoid trailing whitespace.
+  if (!ast.literals || !ast.locations) {
+    throw new Error(
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:105]: Sorry, was waiting to see if you were actually interested in this. I'll go through the coverage and fix it up.

File: lib/compiler/stack.js [hunk 6]
```diff
@@ -22,7 +22,10 @@ class Stack {
     this.ruleName       = ruleName;
     this.type           = type;
     this.bytecode       = bytecode;
-    /* Map from stack index, to label targetting that index */
+    /**
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/stack.js:25]: Yes. This one's pretty easy...

File: lib/peg.d.ts [hunk 7]
```diff
@@ -51,6 +51,23 @@ declare namespace ast {
     location: LocationRange;
   }
 
+  /**
+   * Type of the classes field on a Grammar node. Not quite the same as
+   * CharacterClass (`parts` was renamed to `value`).
+   */
+  interface CharacterRange {
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/peg.d.ts:58]: Will do

File: lib/compiler/passes/generate-js.js [hunk 8]
```diff
@@ -80,9 +85,52 @@ function wrapInSourceNode(prefix, chunk, location, suffix, name) {
...
 }
 
-// Generates parser JavaScript code.
+/**
+ * @typedef {(string|SourceNode)[]} SourceArray
+ *
+ * @typedef {PEG.SourceBuildOptions<PEG.SourceOutputs>} SourceBuildOptions
+ * @typedef {object} ExtraOptions
+ * @property {undefined} [dependencies]
+ * @property {undefined} [exportVar]
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: It's just a typescript trick to make existing code work without adding extra checks. SourceBuildOptions is a union, some elements of which have an optional dependencies, and some don't (similar for exportVar).

But the existing code doesn't care. It says `if (options.dependencies) { ... }`, and if the optional dependencies is there, the clause executes.

But if you type `options` as `SourceBuildOptions`, tsc complains, because `dependencies` might not exist on options. So you either have to add a cast (and I've been trying really hard to avoid those wherever possible - they're ugly), or do some more checking to determine what kind of SourceBuildOptions we've got.

So the trick is to type it as `SourceBuildOptions & ExtraOptions`. This is fine, because anything that can be assigned to `SourceBuildOptions` can be assigned to `SourceBuildOptions & ExtraOptions`, but now I can say `if (options.dependencies) { ... }` again, without typescript complaining.

Basically adding the intersection with ExtraOptions does nothing for the union members where they're already defined and optional (because those fields could already be undefined in those cases), but it allows for them to be set to "undefined" in the other types.

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: TS Playground [link for clarity](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3jAZge0wLhgZxACcBLAOwHMBuGAIwEMi8yBXAW1qiJoYC8B+Zu05EA2gF0aAXyqhIsAEJJUGbHkKlKPRkI5ca6EgPXFyFCdNlzoMAKrKEAHxgKr4GwFEAHsXrK0fIIwLGQAJlCGZFChBkZBIeGR0TBSwMDoIQDGICSYZBj0JAA2+AAULHi2AJSowDD1MCToMOUAdHw1aHUNPQD0vTAA7gAW9CBQAG5c3fWpqelZOXlDmEQA1mUVdjAAZDDevp0zjc1tHSrHfQMjY5PTPXPAQA)

File: lib/compiler/passes/generate-js.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:190]: Actually, I'm wrong. For reasons too subtle for me to comprehend, ts doesn't like `code.map(helper)`, but is happy with `code.map(s => helper(s))`

File: lib/compiler/passes/generate-js.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: But it isn't quite right... (which showed up when I tried to write a test that calls generateJS directly).

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: It needs to be `unknown`, not `undefined` so that you can call it. See [here](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3jAZge0wLhgZxACcBLAOwHMBuGAIwEMi8yBXAW1qiJoYC8B+Zu05EA2gF0aAXyqhIsAEJJUGbHkKlKPRkI5ca6EgPXFyFCdNlzoMAKrKEAHxgKr4GwFEAHsXrK0fIIwLGQA1mSYAO5kBkZBIeFRZDBSwMDoIQDGICSYyej0JAA2+AAULHi2AJSowDD1MCToMOUAdHw1aHUNPQD0vTCRABb0IFAAblzd9amp6Vk5eYOYRKFlFXYwAGQw3r6d043NbR0qh30Dw6MTUz2zaRlk2bn52OWVBz0FxetVsj2RFZrcp-YCpIA)

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: Ugh. I give up.

I've definitely used this previously; and in my case it was important, because the members I was adding had different types in different elements of the union. But here, as you say, dependencies is either defined, and of type `Dependencies`, or it's not defined. And exportVar is either a string, or it's not defined. So we can use those types to make this work.

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:94]: Just for my own sanity, the way I've done this in the past (I went and looked) is to add `{ field?: undefined }` to each member of the union that doesn't already have `field`. Blindly adding it to the whole union doesn't work. Adding `{ field?: unknown }` to the whole union /could/ work, but only if you don't care that when you extract the field from the union, it's typed as `unknown` (but if you extract it from a specific member of the union, its still typed correctly).

File: lib/compiler/stack.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/stack.js:25]: I've fixed it up - would you like me to add it to this diff, or make a separate pull request?

File: lib/compiler/passes/generate-js.js [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// @ts-check
```
[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:1]: I think there's a way to do that regardless.

The problem currently (if I understand correctly) is that `expectType<T>(foo)` is supposed to check that foo is exactly of type T. But there's no way to make that work (with that syntax) in pure typescript. So what typescript actually checks is that `foo extends T`. In particular `expectType<"foo"|"bar">("foo")` passes the typescript check, but fails the tsd check.

But if you're willing to change the syntax a little, and accept slightly less informative error messages, you can get the same checking in pure typescript:

```
function expectExact<T>() {
  return function<U extends T>(
    arg: U
  ): [T] extends [U] ? () => U : never {
    // @ts-expect-error Just for type checking
    return () => arg;
  };
}
```

Now, instead of `expectType<T>(foo)` you have to say `expectExact<T>()(foo)()` (note that its three calls in a row).

With that, you get the exact same error message if `foo` does not extend `T`. But if `T` does not extend `foo`, you get a rather more cryptic:
```
This expression is not callable.
  Type 'never' has no call signatures.
```
Another downside is that if you forget the final `()`, you silently lose the `T extends foo` side of the checking.

On the other hand, these errors would appear in VSCode as you type (and also from the build when you `npm run ts`), rather than only when you `npm test`.

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:1]: In fact, you can do a little better with the error messages:
```
function expectExact<T>() {
  return function<U extends T>(
    tValue: U
  ): [T] extends [U] ? () => U : ["Argument of type", U, "does not match", T] {
    // @ts-expect-error Just for type checking
    return () => tValue;
  };
}
```
Now it reports:
```
This expression is not callable.
  Type '["Argument of type", "foo" | "fuz", "does not match", "bar" | "foo" | "fuz"]' has no call signatures.
```
which isn't great, but does at least tell you what the problem is...

[the remark this answers is not in the document]
Comment (markw65) [reply lib/compiler/passes/generate-js.js:1]: I've put up a [pull request](https://github.com/peggyjs/peggy/pull/436) for discussion.

**Comment by reverofevil** [comment]
JFYI there's a full rewrite in TS [here](https://github.com/reverofevil/peggy/pull/1)

Feel free to steal anything from there, as I don't have time to update it to most recent version

**Comment by markw65** [comment]
@reverofevil Nice. 

From my point of view a full typescript rewrite would be much better. But I'd assumed that sticking with code that "just works" without a build step was probably a project requirement. If thats not the case, then sure...

**Comment by reverofevil** [comment]
@markw65 Whether it's a full rewrite or not, we still have to figure out how to make types work, and it's the same in either case.

As far as I understand, you plan to continue cleaning up peggy's codebase. In that case feel free to take anything from a (differently) cleaned up version of peggy.

**Comment by markw65** [comment]
@hildjj Is there anything else you want here? I have the changes to stack.js if you'd like me to roll those into this pull request; or I can do that separately.

**Comment by markw65** [comment]
@hildjj ping

**Comment by hildjj** [comment]
Sorry, have been traveling. Will get to this when I can, it's still first on the list.

File: lib/compiler/stack.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (hildjj) [reply lib/compiler/stack.js:25]: Let's make it another PR, so we can land this one ASAP.

**Comment by hildjj** [comment]
All this needs is a CHANGELOG.md entry, and it's ready to go.
